File: crates/sargon-uniffi/src/hierarchical_deterministic/bip44/bip44_like_path.rs
use crate::prelude::*;
use sargon::{BIP44LikePath as InternalBIP44LikePath, HasIndex};

/// Either a canonical BIP44 derivation path like so:
///
/// `m / purpose' / coin_type' / account' / change / address_index`
///
/// Or an Radix Olympia BIP44 "like" path, where the `address_index` accidentally
/// was made hardened, i.e.:
///
/// `m / purpose' / coin_type' / account' / change / address_index'`
///
/// This was a mistake made during implementation of Radix Olympia.
///
/// ```
/// extern crate sargon;
/// use sargon::prelude::*;
///
/// fn parse(s: &str) -> Result<BIP44LikePath> {
///    s.parse::<BIP44LikePath>()
/// }
///
/// assert!(parse("m/44'/1022'/0'/0/0").is_ok()); // Canonical BIP44
/// assert!(parse("m/44'/1022'/0'/0/0'").is_ok()); // BIP44 like
///
/// assert_eq!(parse("m/44'/1022'/0'/0'/0"), Err(CommonError::InvalidBIP44LikePathChangeWasUnexpectedlyHardened));
/// assert_eq!(parse("m/44'/1022'/0'/0'/0'"), Err(CommonError::InvalidBIP44LikePathChangeWasUnexpectedlyHardened));
/// assert_eq!(parse("m/44'/0'/0'/0/0'"), Err(CommonError::CoinTypeNotFound { bad_value: 0 }));
/// ```
#[derive(Clone, PartialEq, Eq, Hash, InternalConversion, uniffi::Record)]
pub struct BIP44LikePath {
    pub index: HDPathComponent,
}

#[uniffi::export]
pub fn new_bip44_like_path_from_index(index: HDPathComponent) -> BIP44LikePath {
    InternalBIP44LikePath::new(index.into()).into()
}

#[uniffi::export]
pub fn new_bip44_like_path_from_string(
    string: String,
) -> Result<BIP44LikePath> {
    InternalBIP44LikePath::from_str(&string).into_result()
}

#[uniffi::export]
pub fn bip44_like_path_to_string(path: &BIP44LikePath) -> String {
    path.into_internal().to_string()
}

#[uniffi::export]
pub fn bip44_like_path_get_address_index(
    path: &BIP44LikePath,
) -> HDPathComponent {
    path.into_internal().index().into()
}

#[uniffi::export]
pub fn new_bip44_like_path_sample() -> BIP44LikePath {
    InternalBIP44LikePath::sample().into()
}

#[uniffi::export]
pub fn new_bip44_like_path_sample_other() -> BIP44LikePath {
    InternalBIP44LikePath::sample_other().into()
}

File: jvm/sargon-android/src/main/java/com/radixdlt/sargon/extensions/BIP44LikePath.kt
package com.radixdlt.sargon.extensions

import com.radixdlt.sargon.Bip44LikePath
import com.radixdlt.sargon.DerivationPath
import com.radixdlt.sargon.HdPathComponent
import com.radixdlt.sargon.bip44LikePathGetAddressIndex
import com.radixdlt.sargon.bip44LikePathToString
import com.radixdlt.sargon.newBip44LikePathFromIndex
import com.radixdlt.sargon.newBip44LikePathFromString
import kotlin.jvm.Throws

fun Bip44LikePath.Companion.init(index: HdPathComponent) = newBip44LikePathFromIndex(index = index)

val Bip44LikePath.displayString: String
    get() = bip44LikePathToString(path = this)

val Bip44LikePath.addressIndex: HdPathComponent
    get() = bip44LikePathGetAddressIndex(path = this)

fun Bip44LikePath.asGeneral(): DerivationPath = DerivationPath.Bip44Like(value = this)